processor/groupbytraceprocessor/processor.go

```diff
@@ -0,0 +1,344 @@
+		logger:       logger.With(zap.String("processor", "groupbytrace")),
+		nextConsumer: nextConsumer,
+		config:       config,
+		st:           st,
+		ids:          make([][]byte, config.NumTraces),
+		idToIndex:    make(map[uint64]int),
+		index:        -1, // the first span to be received will be placed at position '0'
+
+		// open the channels
+		removed:  make(chan []byte, 50), // TODO: perf tests to decide what's a good buffer size
+		expired:  make(chan []byte, 50),
+		received: make(chan pdata.Traces, 50),
+		released: make(chan *v1.ResourceSpans, 50),
+
+		// shutdown synchronization
+		stopped:        make(chan context.Context),
+		cancelLock:     &sync.RWMutex{},
+		cancelReleases: false,
+	}
+
+	return sp, nil
+}
+
+func (sp *groupByTraceProcessor) ConsumeTraces(_ context.Context, td pdata.Traces) error {
+	sp.received <- td
+	return nil
+}
+
+func (sp *groupByTraceProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (sp *groupByTraceProcessor) Start(_ context.Context, _ component.Host) error {
+	// starts the processes listening to the channels
+	go sp.onReleased()
+	go sp.onRemoved()
+	go sp.onTrace()
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (sp *groupByTraceProcessor) Shutdown(ctx context.Context) error {
+	sp.stopped <- ctx
+	return nil
+}
+
+func (sp *groupByTraceProcessor) onTrace() {
+	for {
+		// We have a "main loop" here, to avoid real locks in the hot path. The only lock we have
+		// is in the shutdown path, and the processor's state is checked in the funcs that run
+		// in the future.
+		// The other two buffered channels do modifications to the internal map/ring as part of the same thread
+		// as the main loop, so, there's only one access at any given time. Potentially blocking operations
+		// should be done asynchronously, which isn't 100% the case in this initial implementation.
+		select {
+		case ctx := <-sp.stopped:
+			sp.processShutdown(ctx)
+			return
+		case traceID := <-sp.expired:
+			// in this implementation, this won't return errors
+			sp.processExpired(traceID)
+		case item := <-sp.received:
+			// in this implementation, this won't return errors
+			batch := pdata.TracesToOtlp(item)
+			sp.processBatch(batch)
+		}
+	}
+}
+
+func (sp *groupByTraceProcessor) processBatch(batch []*v1.ResourceSpans) error {
+	for _, trace := range batch {
+		// TODO: should we collect all errors?
+		sp.processTrace(trace)
+	}
+
+	return nil
+}
+
+func (sp *groupByTraceProcessor) processTrace(trace *v1.ResourceSpans) error {
+	if trace == nil {
+		// should not happen with the current code
+		return errNilTrace
+	}
+
+	traceID, err := extractTraceID(trace)
+	if err != nil {
+		return fmt.Errorf("couldn't extract trace ID from trace: %w", err)
+	}
+
+	if traceID == nil {
+		return errNilTraceID
+	}
+
+	hashedTraceID := hash(traceID)
```

Comment:
If you keep the `pdata` model instead of converting to OTLP, you will be able to use https://github.com/open-telemetry/opentelemetry-collector/blob/db926fd7f488bfabe0a4558578b2d59c48cc56a8/consumer/pdata/trace.go#L107 and won't need hashing.